src/node/rpc/node_frontend.h

```diff
@@ -409,6 +416,93 @@ namespace ccf
         .set_openapi_hidden(true)
         .install();
 
+      auto promote = [this](auto& args, const nlohmann::json& params) {
+        const auto in = params.get<PromoteNodeToTrusted::In>();
+
+        if (
+          !this->context.get_node_state().is_part_of_network() &&
+          !this->context.get_node_state().is_part_of_public_network() &&
+          !this->context.get_node_state().is_reading_private_ledger())
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "Target node should be part of network to promote nodes");
+        }
+
+        if (!consensus)
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "No configured consensus");
+        }
+
+        if (
+          !this->context.get_node_state().is_primary() &&
+          consensus->type() != ConsensusType::BFT)
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "Only the primary can promote nodes");
+        }
+
+        if (this->context.get_node_state().is_primary())
+        {
+          const auto committed_seqno = consensus->get_committed_seqno_unsafe();
+          const auto node_seqno =
+            consensus->get_confirmed_matching_index(in.node_id);
+
+          if (node_seqno < committed_seqno)
+          {
+            return make_error(
+              HTTP_STATUS_INTERNAL_SERVER_ERROR,
+              ccf::errors::InternalError,
+              "Node has not caught up with primary");
+          }
+        }
+
+        auto service = args.tx.ro(this->network.service);
+        auto active_service = service->get(0);
+
+        if (!active_service.has_value())
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "No service is available to promote node");
+        }
+
+        if (active_service->status != ServiceStatus::OPEN)
+        {
+          return make_error(
+            HTTP_STATUS_BAD_REQUEST,
+            ccf::errors::FrontendNotOpen,
+            "Service must be open to promote nodes");
+        }
+
+        auto nodes = args.tx.rw(this->network.nodes);
+
+        auto node_info = nodes->get(in.node_id);
+        if (!node_info.has_value())
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            fmt::format("Unknown node {}", in.node_id));
+        }
+
+        node_info->status = NodeStatus::TRUSTED;
+        nodes->put(in.node_id, *node_info);
+        return make_success(true);
+      };
+
+      make_endpoint(
+        "promote", HTTP_POST, json_adapter(promote), no_auth_required)
```

Comment:
```suggestion
        "promote", HTTP_POST, json_adapter(promote), {std::make_shared<NodeCertAuthnPolicy>()})
```

Otherwise anyone could promote a node.

Reply:
Great many thanks, I just didn't know what to use here!